diff --git a/frameworks/opengl_wrapper/test/unittest/egl_wrapper_layer_test.cpp b/frameworks/opengl_wrapper/test/unittest/egl_wrapper_layer_test.cpp
index 881767505..ac9e015a7 100644
--- a/frameworks/opengl_wrapper/test/unittest/egl_wrapper_layer_test.cpp
+++ b/frameworks/opengl_wrapper/test/unittest/egl_wrapper_layer_test.cpp
@@ -57,4 +57,58 @@ HWTEST_F(EglWrapperLayerTest, Init002, Level2)
     result = EglWrapperLayer::GetInstance().Init(&dispatchTable);
     ASSERT_TRUE(result);
 }
+
+/**
+ * @tc.name: InitLayers001
+ * @tc.desc:
+ * @tc.type: FUNC
+ */
+HWTEST_F(EglWrapperLayerTest, InitLayers001, Level1)
+{
+    EglWrapperDispatchTable dispatchTable;
+
+    auto result = EglWrapperLayer::GetInstance().InitLayers(&dispatchTable);
+    ASSERT_TRUE(result);
+}
+
+/**
+ * @tc.name: InitLayers002
+ * @tc.desc:
+ * @tc.type: FUNC
+ */
+HWTEST_F(EglWrapperLayerTest, InitLayers002, Level2)
+{
+    EglWrapperDispatchTable dispatchTable;
+    auto result = EglWrapperLayer::GetInstance().InitLayers(&dispatchTable);
+    ASSERT_TRUE(result);
+    result = EglWrapperLayer::GetInstance().InitLayers(&dispatchTable);
+    ASSERT_TRUE(result);
+}
+
+/**
+ * @tc.name: LoadLayers001
+ * @tc.desc:
+ * @tc.type: FUNC
+ */
+HWTEST_F(EglWrapperLayerTest, LoadLayers001, Level1)
+{
+    EglWrapperDispatchTable dispatchTable;
+
+    auto result = EglWrapperLayer::GetInstance().LoadLayers(&dispatchTable);
+    ASSERT_TRUE(result);
+}
+
+/**
+ * @tc.name: LoadLayers002
+ * @tc.desc:
+ * @tc.type: FUNC
+ */
+HWTEST_F(EglWrapperLayerTest, LoadLayers002, Level2)
+{
+    EglWrapperDispatchTable dispatchTable;
+    auto result = EglWrapperLayer::GetInstance().LoadLayers(&dispatchTable);
+    ASSERT_TRUE(result);
+    result = EglWrapperLayer::GetInstance().LoadLayers(&dispatchTable);
+    ASSERT_TRUE(result);
+}
 } // OHOS::Rosen
\ No newline at end of file
